uaclient/entitlements/livepatch.py

```diff
@@ -308,7 +308,7 @@ def process_contract_deltas(
 
         application_status, _ = self.application_status()
         if application_status == status.ApplicationStatus.DISABLED:
-            return True  # only operate on changed directives when ACTIVE
+            return False  # only operate on changed directives when ACTIVE
```

Comment:
@blackboxsw This is the main change I am introducing here, where if the service is marked as disabled, we will mark the `process_contract_deltas` as False. This is how I will be able to detect if a service was successfully enabled during `attach`.

Reply:
@lucasmoura +1 on this changeset, though I think your [service_enabled variable name](https://github.com/canonical/ubuntu-advantage-client/pull/1953/files#diff-50c64ebd349b6ae5bfa48dd3603790d30c8b63141732a0d28ae5aa9fab92b6c5R329) is a misnomer because a True returned from process_contract_deltas can also come in the event that there are [no deltas at all via UAEntitlement.process_contract_deltas](https://github.com/canonical/ubuntu-advantage-client/blob/main/uaclient/entitlements/base.py#L824-L825) regardless of whether the service is truly enabled or disabled.


False retval was originally intended to indicate that no config change operations or deltas were processed. But I think we adapted the base class implementation to return True to allow subclasses to short-circuit and avoid any additional processing within the subclass as there is nothing further/more specific to do.

I think your `return False` from the subclass makes sense given our docstrings and original intent. We may need to tease out how the super().process_contract_deltas behaves in the future for the no delta case, but it really shouldn't make too much difference given that your call-site is really just allowing us to track whether we processed all viable deltas for an enabled service. As long as your event handling isn't broken by a case where no contract deltas exist and process_contract_deltas returns True, then I think your approach is safe.

Reply:
Good point when there is no deltas to apply. I will see if I can address this somehow.
Thanks for that review @blackboxsw

Reply:
@blackboxsw I have updated the code to also check if we have any deltas to apply before marking the service as `processed`